core/transientstore/store.go

```diff
@@ -98,6 +114,173 @@ func (provider *storeProvider) Close() {
+	defer db.Close()
+
+	val, err := db.Get(underDeletionKey)
+	if err != nil {
+		return nil, errors.WithMessagef(err, "retrieving storage marked for deletion")
+	}
+
+	// no storage previously marked as delete
+	if val == nil {
+		return deleteList, nil
+	}
+
+	if err = json.Unmarshal(val, &deleteList); err != nil {
+		return nil, errors.WithMessagef(err, "unmarshalling json delete list: %s", string(val))
+	}
+
+	return deleteList, nil
+}
+
+// Remove a ledger ID from the list of transient storages currently marked for delete.
+func (provider *storeProvider) clearStorageDeletionStatus(ledgerID string) error {
+	dl, err := provider.getStorageMarkedForDeletion()
+	if err != nil {
+		return errors.WithMessagef(err, "clearing storage flag for ledger [%s]", ledgerID)
+	}
+
+	var updatedList []string
+
+	// retain all entries other than the one to be cleared.
+	for i := 0; i < len(dl); i++ {
+		if !strings.EqualFold(ledgerID, dl[i]) {
+			updatedList = append(updatedList, dl[i])
+		}
+	}
+
+	// Nothing to do: ledgerID was not in the current delete list
+	if len(dl) == len(updatedList) {
+		return nil
+	}
+
+	if err = provider.markStorageListForDelete(updatedList); err != nil {
+		return errors.WithMessagef(err, "subtracting [%s] from delete tag list", ledgerID)
+	}
+
+	return nil
+}
+
+// Delete any transient storages that are marked as UNDER_DELETION.  This routine may be called
+// at provider construction to purge any partially deleted transient stores.
+func (provider *storeProvider) processPendingStorageDeletions() error {
+	dl, err := provider.getStorageMarkedForDeletion()
+	if err != nil {
+		return errors.WithMessage(err, "processing pending deletion list")
+	}
+
+	for i := 0; i < len(dl); i++ {
+		err = provider.DeleteStore(dl[i])
+		if err != nil {
+			return errors.WithMessagef(err, "processing delete for storage [%s]", dl[i])
+		}
+
+		err = provider.clearStorageDeletionStatus(dl[i])
+		if err != nil {
+			return errors.WithMessagef(err, "clearing deletion status for storage [%s]", dl[i])
+		}
+	}
+
+	return nil
+}
+
+// Drop removes a transient storage associated with an input channel/ledger.
+// This function must be invoked while the peer is shut down.  To recover from partial deletion
+// due to a crash, the storage will be marked with an UNDER_DELETION status in the a system
+// namespace db.  At the next peer startup, transient storage marked with UNDER_DELETION will
+// be scrubbed from the system.
+func Drop(providerPath, ledgerID string) error {
+	logger.Infof("Dropping ledger [%s] from transient storage path [%s]", ledgerID, providerPath)
+
+	// Ensure the routine is invoked while the peer is down.
+	lockPath := filepath.Join(providerPath, "fileLock")
+	fileLock := leveldbhelper.NewFileLock(lockPath)
+	if err := fileLock.Lock(); err != nil {
+		return errors.WithMessage(err, "as another peer node command is executing,"+
+			" wait for that command to complete its execution or terminate it before retrying")
+	}
```

Comment:
This will require to get this lock at peer start as well - i.e., inside the function `NewStoreProvider`.

Reply:
This was a little annoying but there is a bootstrapping issue trying to construct the provider / lock combination.  My solution was to grab the lock at the start of `Drop` (ensure we're not running at the same time as the peer), immediately release the lock, and make the call to NewProvider, where the lock is re-acquired.   

This opens a few nano/pico seconds of time during which the peer may launch concurrently with three simultaneous lightning strikes - for the rough case of ensuring the peer is down while Dropping the transient storage, it should be OK.

Reply:
A cleaner approach to this could be to have a separate unexported function `newStoreProvider` that assumes that the caller has acquired the lock and the have this call from both the functions `NewStoreProvider` and `Drop` after having acquired the lock.

Reply:
The code was grabbing the lock at the wrong path.  Thanks Manish for your help tracking this down.   The new rev in this PR will set up the file lock at `${transientstore}/../fileLock`